specification/taf-spec.md

```diff
@@ -0,0 +1,915 @@
+outline of the structure for the scripts folder:
+
+```markdown
+/scripts
+    /repo
+      /succeeded - every time a repo is successfully pulled
+      /changed
+      /unchanged
+      /failed - every time a repo is not successfully pulled
+      /completed  - like finally (called in both cases)
+    /update
+      /succeeded - once after all authentication's repositories have been successfully pulled
+      /changed
+      /unchanged
+      /failed - if one repository failed
+      /completed  - like finally (called in both cases)
+```
+
+In TAF's `targets/scripts` directory, there are `update` and `repo` folders that correspond to
+two levels of event triggers. The `repo` folder contains scripts triggered after updating a
+single authentication repository. In contrast, the `update` folder contains scripts that are
+executed once the entire update process for all repositories is complete.
+
+In situations where a TAF authentication repository references other authentication
+repositories, scripts in the repo folder will be executed multiple times. This execution occurs
+for each update of the referenced authentication repositories, as well as for the top-level
+authentication repository itself. On the other hand, scripts in the update folder are triggered
+only once, at the very end of the entire update process.
+
+In TAF, each script is expected to output JSON containing two types of data: persistent and
+transient. The persistent data is automatically saved to a file named `persistent.json` after
+every script execution. This file is then passed on to subsequent scripts. On the other hand,
+transient data, while also passed to the next script, is not stored permanently.
+In addition to handling persistent and transient data, these scripts also receive detailed
+information about the repositories involved — this includes both the authentication repository
+and its target repositories.
+
+### 5.1.4 Updating repositories with cross-references
+
+Within TAF, an authentication repository has the capability to reference other authentication
+repositories. This relationship is defined in the `dependencies.json` file. Alongside the name of
+each referenced repository, the file can also specify out-of-band commits. The URLs for these
+referenced authentication repositories are determined similarly to target repositories. This involves
+combining their full names in a `namespace/name`format with URL templates specified in `mirrors.
+json`.
+
+Once the primary authentication repository (including its own contents and targets) has been
+validated without errors, the updater processes each repository listed in the `dependencies.json`. It
+determines their URLs and then recursively updates each referenced authentication repository. When an
+out-of-band commit is specified for a referenced repository, it is included in the update function.
+The updater then conducts a validation check based on this commit, as described earlier.
+
```

Comment:
Now that I've read the 5. Updater section, it does feel like we can move everything that we've written here to the [Updater ](https://github.com/openlawlibrary/taf/blob/master/docs/updater/update_process.md?plain=1#L36) docs, and then reference here in the spec. That would simplify the specification, but also leave it up to the reader who's interested in more details to go over? What are your thoughts?

Reply:
I think it's fine to reference other documents if there are extensions of the specification, but it should contain everything that describes TAF main functionalities. I don't think that descriptions of how to call the CLI command should go into the specification, but it should be possible to understand how it work. I think that we should figure out if there is anything else that is in these other docs that we should mention here. The old updater specification is very outdated and not needed now that we have this, so I will remove it. The other docs should be focused on how to run these commands, which I think is the case for the most part. Maybe we can reference the specification from other docs instead? Although I do not think that it's necessary to update other docs as a part of this PR